### PR TITLE
Add whitespace before markers to make editable package can be installed by pip

### PR DESCRIPTION
**Pull request by jxltom**
Fixed https://github.com/sarugaku/requirementslib/issues/70

### CODE REVIEW

**Comment by uranusjr** [comment]
I wonder if we should just modify how parts are joined (`Requirement.as_line()`) to join with a whilespece instead of `""`.

https://github.com/sarugaku/requirementslib/blob/d457054d969fcb5e712c8f74f9ddac182335443b/src/requirementslib/models/requirements.py#L950-L958

**Comment by jxltom** [comment]
I don't have a preference. :D
Any recommendations?

**Comment by uranusjr** [comment]
I think @techalchemy is best to answer this

File: src/requirementslib/models/requirements.py [hunk 1]
```diff
@@ -760,7 +760,7 @@ def hashes_as_pip(self):
     @property
     def markers_as_pip(self):
         if self.markers:
-            return "; {0}".format(self.markers).replace('"', "'")
+            return " ; {0}".format(self.markers).replace('"', "'")
```
Comment (techalchemy) [inline src/requirementslib/models/requirements.py]: Is this real... this definitely isn't how it used to be parsed... that is super annooying

**Comment by techalchemy** [comment]
>I wonder if we should just modify how parts are joined (Requirement.as_line()) to join with a whilespece instead of "".

We can't actually because you will end up with a space between the requirement and extras,  which is invalid obviously.

What's weird is that the line parser in pip is still:

```python
    if is_url(name):
        marker_sep = '; '
    else:
        marker_sep = ';'
```

So possibly the correct answer is that we shouldn't have any spaces at all unless it's a URL... can you check if that works?

**Comment by uranusjr** [comment]
> a space between the requirement and extras, which is invalid obviously.

Not really :p

https://www.python.org/dev/peps/pep-0508/#examples

I believe spaces between parts are always allowed, but *required* if the part is a path or URL (since you can have semicolons in them).

The PEP also contains a [grammar walk-down](https://www.python.org/dev/peps/pep-0508/#grammar). As it listed at the end of the section, whitespaces can occur (and must after a URL if markers follow) between all parts.

(As noted in the PEP, pip does not implement all grammars, but it does regarding whitespaces.)

**Comment by techalchemy** [comment]
> I believe spaces between parts are always allowed, but required if the part is a path or URL (since you can have semicolons in them).

Yes, but note that my comment said _extras_, which are also joined in the same statement, and which are not markers.


> The PEP also contains a grammar walk-down. As it listed at the end of the section, whitespaces can occur (and must after a URL if markers follow) between all parts.

_markers_ must have a space for urls, which I know and which is why I just put a space all the time (because I did read the grammar and I already knew this worked fine).  IF you put a space between every aspect we join, however, you will then join in the URL + Extras string (which is part of the egg fragment) with an additional unescaped space which will create possibly unexpected outcomes.

I should also point out that this is completely irrelevant because the output is passed to _pip_ which as you mentioned doesn't actually contain a full implementation of the grammar, which is the entire point of this library acting as an interface.  The expected input for a urlspec as compared with the input required to generate an `InstallRequirement` are completely different.

**Comment by techalchemy** [comment]
For instance:

```pycon
>>> plette = InstallRequirement.from_line("https://github.com/sarugaku/plette/archive/0.2.2.tar.gz#egg=plette[validation]; python_version <= '3.5'")
>>> plette
<InstallRequirement object: plette[validation] from https://github.com/sarugaku/plette/archive/0.2.2.tar.gz#egg=plette[validation] editable=False>
>>> plette.markers
<Marker('python_version <= "3.5"')>
```

And the actual grammar implementation given the same input:

```pycon
>>> pkg_resources.Requirement.parse("https://github.com/sarugaku/plette/archive/0.2.2.tar.gz#egg=plette[validation]; python_version <= '3.5'")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hawk/.pyenv/versions/3.7.0/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3002, in parse
    req, = parse_requirements(s)
  File "/home/hawk/.pyenv/versions/3.7.0/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2949, in parse_requirements
    yield Requirement(line)
  File "/home/hawk/.pyenv/versions/3.7.0/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2958, in __init__
    raise RequirementParseError(str(e))
pkg_resources.RequirementParseError: Invalid requirement, parse error at "'://githu'"
```

Because you actually have to normalize that input to be this:

```pycon
>>> pkg_resources.Requirement.parse("plette[validation] @ https://github.com/sarugaku/plette/archive/0.2.2.tar.gz ; python_version <= '3.5'")
Requirement.parse('plette[validation]@ https://github.com/sarugaku/plette/archive/0.2.2.tar.gz; python_version <= "3.5"')
```

As you can see even if you do put an extra space, the "correct" syntax has no space between the url and the semicolon, or the name+extras and the `@`.  Pip is not going to fix parsing logic because they have a standardized user input and they reformat that input to suit the PEP after parsing it.  If there is a bug, that is another question, but they are not going to break things as they already know about this already, they do direct conversions to `pkg_resources.Requirement` objects just like I am doing.

**Comment by techalchemy** [comment]
Nothing, like I told you already, besides the number of libraries separating us from the actual grammar implementation.  All it takes is one library deciding to handle it differently and considering the case involves URLs, that is a pretty easy possibility -- say someone urlencodes spaces, for example.  We did that ourselves. Why would you want to add an extra space when the accurate representation by the reference implementation spits it back out with no spaces anyway? you're literally just putting a character so that it can be stripped. That is what you're advocating. Why?

**Comment by techalchemy** [comment]
Thanks for the PR, merging as is

**Pull request merged**